File: webagent-ui/src/lib/clawd/agent-turns.ts
export type AgentTurnStatus = "queued" | "running" | "succeeded" | "interrupted" | "failed";

export interface AgentTurnStep {
  id: string;
  kind: "retrieval" | "tool" | "expert" | "citation" | "artifact" | "generation";
  label: string;
  detail: string | null;
  status: "running" | "succeeded" | "failed" | "skipped";
  started_at_ms: number | null;
  completed_at_ms: number | null;
  public_payload: unknown;
  debug_payload?: unknown;
}

export interface AgentCitation {
  id: string;
  number: number;
  source_kind: string;
  source_label: string;
  title: string | null;
  location: string | null;
  preview: string;
  debug_payload?: unknown;
}

export interface AgentExpertResult {
  expert_name: string;
  status: "running" | "succeeded" | "retrying" | "failed" | "skipped";
  summary: string | null;
  citation_numbers: number[];
  error: string | null;
}

export interface AgentTurnRecord {
  id: string;
  conversation_id: string;
  tenant_id: string | null;
  owner_id: string;
  user_message: string;
  assistant_text: string;
  status: AgentTurnStatus;
  started_at_ms: number;
  completed_at_ms: number | null;
  steps: AgentTurnStep[];
  citations: AgentCitation[];
  expert_results: AgentExpertResult[];
  artifacts: unknown[];
  error: { public_message: string; debug_message: string | null; code: string | null } | null;
  debug_events: Array<{ event_type: string; at_ms: number; payload: unknown }>;
}

export type AgentPipelineGroupKind =
  | "retrieval"
  | "tool"
  | "expert"
  | "artifact"
  | "generation";

export interface AgentPipelineItem {
  id: string;
  title: string;
  action: string;
  output: string;
  status: AgentTurnStep["status"] | AgentExpertResult["status"];
  references: number[];
  detail: string | null;
  started_at_ms: number | null;
  completed_at_ms: number | null;
  children?: AgentPipelineItem[];
}

export interface AgentPipelineGroup {
  kind: AgentPipelineGroupKind;
  title: string;
  items: AgentPipelineItem[];
}

const PIPELINE_GROUP_TITLES: Record<AgentPipelineGroupKind, string> = {
  retrieval: "资料检索",
  tool: "工具执行",
  expert: "专家分析",
  artifact: "产物生成",
  generation: "回答生成",
};

const PIPELINE_GROUP_ORDER: AgentPipelineGroupKind[] = [
  "generation",
  "tool",
  "retrieval",
  "expert",
  "artifact",
];

export function groupAgentTurnSteps(
  steps: AgentTurnStep[],
  citations: AgentCitation[],
  expertResults: AgentExpertResult[],
): AgentPipelineGroup[] {
  const groups = new Map<AgentPipelineGroupKind, AgentPipelineGroup>();
  const byId = new Map<string, AgentPipelineItem>();
  let lastToolParentId: string | null = null;

  const attachToParent = (item: AgentPipelineItem, parentId: string | null): void => {
    const resolvedParentId = parentId && byId.has(parentId) ? parentId : lastToolParentId;
    if (!resolvedParentId) {
      addPipelineItem(groups, pipelineGroupKindForItem(item), item);
      byId.set(item.id, item);
      if (item.action === "工具调用" || item.title.includes("工具")) {
        lastToolParentId = item.id;
      }
      return;
    }

    const parent = byId.get(resolvedParentId);
    if (!parent) {
      addPipelineItem(groups, pipelineGroupKindForItem(item), item);
      byId.set(item.id, item);
      return;
    }

    parent.children = [...(parent.children ?? []), item];
    byId.set(item.id, item);
  };

  for (const step of steps) {
    const kind = pipelineGroupKindForStep(step.kind);
    const item: AgentPipelineItem = {
      id: step.id,
      title: step.label,
      action: actionForStep(step),
      output: outputForStep(step),
      status: step.status,
      references: citationNumbersFromPayload(step.public_payload),
      detail: step.detail,
      started_at_ms: step.started_at_ms,
      completed_at_ms: step.completed_at_ms,
    };
    const payload = payloadRecord(step.public_payload);
    const parentId = stringPayload(payload, "parent_id") ?? stringPayload(payload, "parentId");
    if (kind === "tool" && isToolParentItem(item)) {
      lastToolParentId = item.id;
      addPipelineItem(groups, kind, item);
      byId.set(item.id, item);
      continue;
    }

    if (kind === "generation" && isTopLevelGenerationItem(item)) {
      addPipelineItem(groups, kind, item);
      byId.set(item.id, item);
      continue;
    }

    if (parentId || lastToolParentId) {
      attachToParent(item, parentId);
    } else {
      addPipelineItem(groups, kind, item);
      byId.set(item.id, item);
    }
  }

  expertResults.forEach((result, index) => {
    const item: AgentPipelineItem = {
      id: `expert-result-${result.expert_name}-${index}`,
      title: `${result.expert_name} 分析`,
      action: "专家视角分析",
      output: result.summary ?? result.error ?? "专家分析已更新",
      status: result.status,
      references: result.citation_numbers,
      detail: null,
      started_at_ms: null,
      completed_at_ms: null,
    };
    if (lastToolParentId && byId.has(lastToolParentId)) {
      const parent = byId.get(lastToolParentId);
      if (parent) {
        parent.children = [...(parent.children ?? []), item];
        return;
      }
    }
    addPipelineItem(groups, "expert", item);
  });

  if (!steps.length && citations.length) {
    addPipelineItem(groups, "retrieval", {
      id: "citations-summary",
      title: "引用资料",
      action: "资料检索",
      output: `形成 ${citations.length} 条引用`,
      status: "succeeded",
      references: citations.map((citation) => citation.number),
      detail: null,
      started_at_ms: null,
      completed_at_ms: null,
    });
  }

  return PIPELINE_GROUP_ORDER.flatMap((kind) => {
    const group = groups.get(kind);
    return group && group.items.length ? [group] : [];
  });
}

export function statusLabelForAgentTurn(status: AgentTurnStatus): string {
  switch (status) {
    case "queued":
      return "等待发送";
    case "running":
      return "正在处理";
    case "succeeded":
      return "已完成";
    case "interrupted":
      return "已停止";
    case "failed":
      return "处理失败";
  }
}

function addPipelineItem(
  groups: Map<AgentPipelineGroupKind, AgentPipelineGroup>,
  kind: AgentPipelineGroupKind,
  item: AgentPipelineItem,
): void {
  const group = groups.get(kind) ?? {
    kind,
    title: PIPELINE_GROUP_TITLES[kind],
    items: [],
  };
  group.items.push(item);
  groups.set(kind, group);
}

function isToolParentItem(item: AgentPipelineItem): boolean {
  return item.title.includes("工具") || item.action.includes("调用") || item.action === "工具调用";
}

function isTopLevelGenerationItem(item: AgentPipelineItem): boolean {
  return item.title.includes("模型") || item.title.includes("计划") || item.title.includes("响应");
}

function pipelineGroupKindForItem(item: AgentPipelineItem): AgentPipelineGroupKind {
  if (item.title.includes("模型") || item.title.includes("计划") || item.title.includes("响应")) {
    return "generation";
  }
  if (item.action.includes("调用") || item.title.includes("工具")) {
    return "tool";
  }
  if (item.title.includes("专家")) {
    return "expert";
  }
  return "retrieval";
}

function pipelineGroupKindForStep(kind: AgentTurnStep["kind"]): AgentPipelineGroupKind {
  switch (kind) {
    case "retrieval":
    case "citation":
      return "retrieval";
    case "expert":
      return "expert";
    case "artifact":
      return "artifact";
    case "generation":
      return "generation";
    case "tool":
      return "tool";
  }
}

function actionForStep(step: AgentTurnStep): string {
  const payload = payloadRecord(step.public_payload);
  const query = stringPayload(payload, "query");
  if (query) return `查询：${query}`;

  const sourceName = stringPayload(payload, "source_name") ?? stringPayload(payload, "sourceName");
  if (sourceName) return `来源：${sourceName}`;

  return step.label;
}

function outputForStep(step: AgentTurnStep): string {
  const payload = payloadRecord(step.public_payload);
  const resultSummary = stringPayload(payload, "result_summary");
  if (resultSummary) return resultSummary;

  const hitCount = numberPayload(payload, "hit_count");
  if (hitCount !== null) {
    const references = numberListPayload(payload, "citation_numbers");
    const referenceSummary = references.length ? `，引用 ${formatCitationNumbers(references)}` : "";
    return `命中 ${hitCount} 篇资料${referenceSummary}`;
  }

  return step.detail ?? "步骤已更新";
}

function citationNumbersFromPayload(payload: unknown): number[] {
  return numberListPayload(payloadRecord(payload), "citation_numbers");
}

function formatCitationNumbers(numbers: number[]): string {
  return numbers.map((number) => `[${number}]`).join(" ");
}

export function payloadRecord(payload: unknown): Record<string, unknown> {
  if (payload && typeof payload === "object" && !Array.isArray(payload)) {
    return payload as Record<string, unknown>;
  }
  return {};
}

export function stringPayload(payload: Record<string, unknown>, key: string): string | null {
  return stringValue(payload[key]);
}

export function numberPayload(payload: Record<string, unknown>, key: string): number | null {
  return numericValue(payload[key]);
}

export function numberListPayload(payload: Record<string, unknown>, key: string): number[] {
  const value = payload[key];
  if (!Array.isArray(value)) return [];

  return value.filter((item): item is number => Number.isFinite(item));
}

function stringValue(value: unknown): string | null {
  return typeof value === "string" && value.trim() ? value : null;
}

function numericValue(value: unknown): number | null {
  return typeof value === "number" && Number.isFinite(value) ? value : null;
}

export function replaceEvidenceMarkersWithCitationNumbers(
  text: string,
  citations: Array<{ id: string; number: number }>,
): string {
  return text.replace(/evidence:([^\s`),.;，。；）]+)/g, (raw, id: string) => {
    const citation = citations.find((item) => item.id === id);
    return citation ? `[${citation.number}]` : raw;
  });
}

export function groupTurnsByDisplayDate<T extends { started_at_ms: number }>(
  turns: T[],
  now: Date = new Date(),
): Array<{ key: string; label: string; turns: T[] }> {
  const groups = new Map<string, { key: string; label: string; turns: T[] }>();
  for (const turn of turns) {
    const date = new Date(turn.started_at_ms);
    const key = date.toLocaleDateString("zh-CN");
    const sameYear = date.getFullYear() === now.getFullYear();
    const label = formatDisplayDate(date, sameYear);
    const group = groups.get(key) ?? { key, label, turns: [] };
    group.turns.push(turn);
    groups.set(key, group);
  }
  return [...groups.values()];
}

const WEEKDAYS = ["周日", "周一", "周二", "周三", "周四", "周五", "周六"];

function formatDisplayDate(date: Date, sameYear: boolean): string {
  const monthDay = `${date.getMonth() + 1}月${date.getDate()}日`;
  const weekday = WEEKDAYS[date.getDay()];
  return sameYear ? `${monthDay} ${weekday}` : `${date.getFullYear()}年${monthDay} ${weekday}`;
}

File: rust/crates/clawd/src/agent_turns.rs
use serde::{Deserialize, Serialize};
use serde_json::Value;

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq, Eq)]
#[serde(rename_all = "snake_case")]
pub enum AgentConversationStatus {
    Idle,
    Running,
    Interrupted,
    Failed,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct AgentConversationRecord {
    pub id: String,
    pub tenant_id: Option<String>,
    pub owner_id: String,
    pub title: String,
    pub status: AgentConversationStatus,
    pub selected_knowledge_base_ids: Vec<String>,
    pub selected_data_source_ids: Vec<String>,
    pub selected_expert_ids: Vec<String>,
    pub model_profile_id: Option<String>,
    pub created_at_ms: u64,
    pub updated_at_ms: u64,
}

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq, Eq)]
#[serde(rename_all = "snake_case")]
pub enum AgentTurnStatus {
    Queued,
    Running,
    Succeeded,
    Interrupted,
    Failed,
}

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq, Eq)]
#[serde(rename_all = "snake_case")]
pub enum AgentTurnStepKind {
    Retrieval,
    Tool,
    Expert,
    Citation,
    Artifact,
    Generation,
}

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq, Eq)]
#[serde(rename_all = "snake_case")]
pub enum AgentTurnStepStatus {
    Running,
    Succeeded,
    Failed,
    Skipped,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct AgentTurnStep {
    pub id: String,
    pub kind: AgentTurnStepKind,
    pub label: String,
    pub detail: Option<String>,
    pub status: AgentTurnStepStatus,
    pub started_at_ms: Option<u64>,
    pub completed_at_ms: Option<u64>,
    pub public_payload: Value,
    pub debug_payload: Option<Value>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct AgentCitation {
    pub id: String,
    pub number: u32,
    pub source_kind: String,
    pub source_label: String,
    pub title: Option<String>,
    pub location: Option<String>,
    pub preview: String,
    pub debug_payload: Option<Value>,
}

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq, Eq)]
#[serde(rename_all = "snake_case")]
pub enum AgentExpertStatus {
    Running,
    Succeeded,
    Retrying,
    Failed,
    Skipped,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct AgentExpertResult {
    pub expert_name: String,
    pub status: AgentExpertStatus,
    pub summary: Option<String>,
    pub citation_numbers: Vec<u32>,
    pub error: Option<String>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct AgentTurnError {
    pub public_message: String,
    pub debug_message: Option<String>,
    pub code: Option<String>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct AgentTurnDebugEvent {
    pub event_type: String,
    pub at_ms: u64,
    pub payload: Value,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct AgentTurnRecord {
    pub id: String,
    pub conversation_id: String,
    pub tenant_id: Option<String>,
    pub owner_id: String,
    pub user_message: String,
    pub assistant_text: String,
    pub status: AgentTurnStatus,
    pub started_at_ms: u64,
    pub completed_at_ms: Option<u64>,
    pub steps: Vec<AgentTurnStep>,
    pub citations: Vec<AgentCitation>,
    pub expert_results: Vec<AgentExpertResult>,
    pub artifacts: Vec<Value>,
    pub error: Option<AgentTurnError>,
    pub debug_events: Vec<AgentTurnDebugEvent>,
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn agent_turn_round_trips_structured_steps_and_citations() {
        let turn = AgentTurnRecord {
            id: "turn-1".to_string(),
            conversation_id: "conv-1".to_string(),
            tenant_id: Some("tenant-a".to_string()),
            owner_id: "alice".to_string(),
            user_message: "分析台海供应链风险".to_string(),
            assistant_text: "主要风险来自运输、制裁和关键零部件。[1]".to_string(),
            status: AgentTurnStatus::Succeeded,
            started_at_ms: 1_700_000_000_000,
            completed_at_ms: Some(1_700_000_001_000),
            steps: vec![AgentTurnStep {
                id: "step-1".to_string(),
                kind: AgentTurnStepKind::Retrieval,
                label: "检索平台资料".to_string(),
                detail: Some("命中 3 篇文档".to_string()),
                status: AgentTurnStepStatus::Succeeded,
                started_at_ms: Some(1_700_000_000_100),
                completed_at_ms: Some(1_700_000_000_600),
                public_payload: serde_json::json!({ "hit_count": 3 }),
                debug_payload: Some(serde_json::json!({ "tool": "EsSearch" })),
            }],
            citations: vec![AgentCitation {
                id: "cite-1".to_string(),
                number: 1,
                source_kind: "es".to_string(),
                source_label: "平台资料库".to_string(),
                title: Some("供应链风险报告".to_string()),
                location: Some("index-a#hit-1".to_string()),
                preview: "关键港口拥堵会放大风险。".to_string(),
                debug_payload: Some(serde_json::json!({ "score": 0.91 })),
            }],
            expert_results: vec![AgentExpertResult {
                expert_name: "Howard-Wang".to_string(),
                status: AgentExpertStatus::Succeeded,
                summary: Some("PLA logistics indicators require separate tracking.".to_string()),
                citation_numbers: vec![1],
                error: None,
            }],
            artifacts: Vec::new(),
            error: None,
            debug_events: Vec::new(),
        };

        let raw = serde_json::to_string(&turn).expect("serialize turn");
        let parsed: AgentTurnRecord = serde_json::from_str(&raw).expect("deserialize turn");

        assert_eq!(parsed.id, "turn-1");
        assert_eq!(parsed.steps[0].kind, AgentTurnStepKind::Retrieval);
        assert_eq!(parsed.citations[0].number, 1);
        assert_eq!(parsed.expert_results[0].expert_name, "Howard-Wang");
    }
}
